### PR TITLE
KnowledgeBase v2

### PR DESCRIPTION
**Pull request by ltfish**
While angr's knowledge base (the `KnowledgeBase` class) has existed for a long time, it was never used to its full extent. One notorious example is `CFGFast` (which is itself an analysis) where all artifacts are simply stored inside multiple dicts under the class. As a result, dumping (pickling) a CFG is unnecessarily complex and there is no guarantee that anything should have worked. Other consequences include the myth of importing CFG from or exporting artifacts to another tool (IDA Pro, Binary Ninja, or radare2). Obviously, CFG isn't the only victim -- you cannot even easily export a `Function` object unless you write some non-trivial serialization code to do so despite the fact that `Function` itself is designed and regarded as a first-class citizen in angr's knowledge base design.

As far as I see it, a knowledge base in angr is a collection of consistent facts and artifacts (which are, knowledge). It should be the foundation of data exchange and sharing.

This PR aims to do the following things:

- Splitting data/storage from the logic code in angr's analysis. For now, CFGFast is done: It is split into a `CFGModel` (which is both a model and a view into the model) and the analysis code (still in `angr.analyses.cfg.cfg_fast`). This is close to an MVC pattern.
- Make artifacts in knowledge base serializable. Now you can serialize (and unserialize) basic artifacts in knowledge base into a compact binary representation (thanks to protobuf). Hopefully this makes it easier to share data between angr and other tools.
- Deprecate the `object` notion in `KnowledgeBase`. angr's knowledge base targets an entire project, not just a single binary.

Of course, there are many TODOs:
- [ ] Serialize calling convention and function prototypes.
- [ ] Test dumping CFG from IDA Pro and Binary Ninja and importing the serialized data into angr.
- [ ] Make other analyses MVC as well.
- [ ] Introduce analysis dependencies: An analysis should not depend on another analysis. Rather, it should depend on the existence of another model/artifact. If the model/artifact does not exist, it may invoke another analysis that produces the model/artifact with a default configuration.

Related issues: #1444, #1484

### CODE REVIEW

**Comment by ltfish** [comment]
I will merge this branch in as soon as tests pass so we can all work on the latest code base. All TODOs will be resolved soon after.

**Comment by zardus** [comment]
I agree 100%. Merge merge merge!

File: angr/analyses/cfg/cfg_fast.py [hunk 1]
```diff
@@ -1113,13 +665,33 @@ def _calc_entropy(data, size=None):
...
-        return self.insn_addr_to_memory_data
+        return self._model.insn_addr_to_memory_data
+
+    @property
+    def _memory_data(self):
+        return self._model.memory_data
+
+    @property
+    def memory_data(self):
+        return self._model.memory_data
```
Comment (cq674350529) [inline angr/analyses/cfg/cfg_fast.py]: just keep one ?

File: angr/analyses/cfg/cfg_fast.py [hunk 2]
```diff
@@ -1113,13 +665,33 @@ def _calc_entropy(data, size=None):
...
         l.warning('_insn_addr_to_memory_data has been made public and is deprecated. Please fix your code accordingly.')
-        return self.insn_addr_to_memory_data
+        return self._model.insn_addr_to_memory_data
+
+    @property
+    def _memory_data(self):
+        return self._model.memory_data
+
+    @property
+    def memory_data(self):
```
Comment (cq674350529) [inline angr/analyses/cfg/cfg_fast.py]: just keep one?

File: angr/analyses/cfg/cfg_fast.py [hunk 3]
```diff
@@ -1113,13 +665,33 @@ def _calc_entropy(data, size=None):
...
+        return self._model.graph
 
     @property
     def _insn_addr_to_memory_data(self):
         l.warning('_insn_addr_to_memory_data has been made public and is deprecated. Please fix your code accordingly.')
-        return self.insn_addr_to_memory_data
+        return self._model.insn_addr_to_memory_data
+
+    @property
+    def _memory_data(self):
```
Comment (cq674350529) [inline angr/analyses/cfg/cfg_fast.py]: just keep one?

File: angr/analyses/cfg/cfg_fast.py [hunk 4]
```diff
@@ -1113,13 +665,33 @@ def _calc_entropy(data, size=None):
...
+        return self._model.memory_data
+
```
Comment (cq674350529) [inline angr/analyses/cfg/cfg_fast.py]: just keep only one?

File: angr/analyses/cfg/cfg_fast_soot.py [hunk 5]
```diff
@@ -118,15 +118,15 @@ def _generate_cfgnode(self, cfg_job, current_function_addr):
 
         try:
 
-            if addr in self._nodes:
-                cfg_node = self._nodes[addr]
+            if addr in self.model.nodes():
+                cfg_node = self.model.nodes()[addr]
                 soot_block = cfg_node.soot_block
```
Comment (cq674350529) [inline angr/analyses/cfg/cfg_fast_soot.py]: Wrong behaviour? for `self.model.nodes()` returns "an iterator of all nodes in the graph".

File: angr/engines/vex/engine.py [hunk 6]
```diff
@@ -241,7 +242,8 @@ def _handle_irsb(self, state, successors, irsb, skip_stmts, last_stmt, whitelist
 
         # if we've told the block to truncate before it ends, it will definitely have a default
         # exit barring errors
-        has_default_exit = has_default_exit and (last_stmt in (None, 'default') or num_stmts <= last_stmt)
+        has_default_exit = has_default_exit and (last_stmt in (None, 'default', DEFAULT_STATEMENT) or
```
Comment (cq674350529) [inline angr/engines/vex/engine.py]: remove 'default' instead?

File: angr/engines/vex/engine.py [hunk 7]
```diff
@@ -263,7 +265,8 @@ def _handle_irsb(self, state, successors, irsb, skip_stmts, last_stmt, whitelist
             if stmt_idx < skip_stmts:
                 l.debug("Skipping statement %d", stmt_idx)
                 continue
-            if last_stmt is not None and last_stmt != 'default' and stmt_idx > last_stmt:
+            if last_stmt is not None and last_stmt != 'default' and last_stmt != DEFAULT_STATEMENT \
```
Comment (cq674350529) [inline angr/engines/vex/engine.py]: remove `last_stmt != 'default'` instead?

File: angr/analyses/cfg/cfg_fast.py (hunk 4 above)
Comment (ltfish) [reply angr/analyses/cfg/cfg_fast.py]: I'm keeping both `_memory_data` and `memory_data` for compatibility concerns.

File: angr/analyses/cfg/cfg_fast_soot.py (hunk 5 above)
Comment (ltfish) [reply angr/analyses/cfg/cfg_fast_soot.py]: Good catch!

File: angr/engines/vex/engine.py (hunk 6 above)
Comment (ltfish) [reply angr/engines/vex/engine.py]: Good catch!

File: angr/engines/vex/engine.py (hunk 7 above)
Comment (ltfish) [reply angr/engines/vex/engine.py]: Good catch!

**Comment by sraboy** [comment]
Any docs on breaking changes?

**Comment by ltfish** [comment]
@sraboy As far as I know there are no breaking changes besides a few internal data structures (e.g., `MemoryData`). All old APIs still exist. Some of them are marked as deprecated and will print out deprecation warnings with new API suggestions when used. All our tests pass. Let me know if you spot anything breaking!

Oh, you do need to install protobuf in your virtualenv: `pip install protobuf`.

**Pull request merged**